include/gbinder_types.h

```diff
@@ -167,16 +167,66 @@ GBinderLocalReply*
 
 #define GBINDER_TX_FLAG_ONEWAY (0x01)
 
+/* other possible errors from system/core/libutils/include/utils/Errors.h */
 typedef enum gbinder_status {
     GBINDER_STATUS_OK = 0,
     GBINDER_STATUS_FAILED,
-    GBINDER_STATUS_DEAD_OBJECT
+    GBINDER_STATUS_UNKNOWN_ERROR       = (-2147483647-1), // INT32_MIN value
+    GBINDER_STATUS_NO_MEMORY           = -ENOMEM,
+    GBINDER_STATUS_INVALID_OPERATION   = -ENOSYS,
+    GBINDER_STATUS_BAD_VALUE           = -EINVAL,
+    GBINDER_STATUS_BAD_TYPE            = (GBINDER_STATUS_UNKNOWN_ERROR + 1),
+    GBINDER_STATUS_NAME_NOT_FOUND      = -ENOENT,
+    GBINDER_STATUS_PERMISSION_DENIED   = -EPERM,
+    GBINDER_STATUS_NO_INIT             = -ENODEV,
+    GBINDER_STATUS_ALREADY_EXISTS      = -EEXIST,
+    GBINDER_STATUS_DEAD_OBJECT         = -EPIPE,
+    GBINDER_STATUS_FAILED_TRANSACTION  = (GBINDER_STATUS_UNKNOWN_ERROR + 2),
+#if !defined(_WIN32)
+    GBINDER_STATUS_BAD_INDEX           = -EOVERFLOW,
+    GBINDER_STATUS_NOT_ENOUGH_DATA     = -ENODATA,
+    GBINDER_STATUS_WOULD_BLOCK         = -EWOULDBLOCK, 
+    GBINDER_STATUS_TIMED_OUT           = -ETIMEDOUT,
+    GBINDER_STATUS_UNKNOWN_TRANSACTION = -EBADMSG,
+#else
+    GBINDER_STATUS_BAD_INDEX           = -E2BIG,
+    GBINDER_STATUS_NOT_ENOUGH_DATA     = (GBINDER_STATUS_UNKNOWN_ERROR + 3),
+    GBINDER_STATUS_WOULD_BLOCK         = (GBINDER_STATUS_UNKNOWN_ERROR + 4),
+    GBINDER_STATUS_TIMED_OUT           = (GBINDER_STATUS_UNKNOWN_ERROR + 5),
+    GBINDER_STATUS_UNKNOWN_TRANSACTION = (GBINDER_STATUS_UNKNOWN_ERROR + 6),
+#endif    
+    GBINDER_STATUS_FDS_NOT_ALLOWED     = (GBINDER_STATUS_UNKNOWN_ERROR + 7),
+    GBINDER_STATUS_UNEXPECTED_NULL     = (GBINDER_STATUS_UNKNOWN_ERROR + 8),
 } GBINDER_STATUS;
 
+/* frameworks/native/libs/binder/include/binder/IBinder.h */
+#define GBINDER_FIRST_CALL_TRANSACTION    (0x00000001)
 #define GBINDER_FOURCC(c1,c2,c3,c4) \
     (((c1) << 24) | ((c2) << 16) | ((c3) << 8) | (c4))
-
-#define GBINDER_FIRST_CALL_TRANSACTION    (0x00000001)
+#define GBINDER_TRANSACTION(c2,c3,c4)       GBINDER_FOURCC('_',c2,c3,c4)
+#define GBINDER_PING_TRANSACTION            GBINDER_TRANSACTION('P','N','G')
+#define GBINDER_DUMP_TRANSACTION            GBINDER_TRANSACTION('D','M','P')
+#define GBINDER_SHELL_COMMAND_TRANSACTION   GBINDER_TRANSACTION('C','M','D')
+#define GBINDER_INTERFACE_TRANSACTION       GBINDER_TRANSACTION('N','T','F')
+#define GBINDER_SYSPROPS_TRANSACTION        GBINDER_TRANSACTION('S','P','R')
+#define GBINDER_EXTENSION_TRANSACTION       GBINDER_TRANSACTION('E','X','T'),
+#define GBINDER_DEBUG_PID_TRANSACTION       GBINDER_TRANSACTION('P','I','D'),
+#define GBINDER_SET_RPC_CLIENT_TRANSACTION  GBINDER_TRANSACTION('R','P','C'),
+#undef GBINDER_FOURCC
```

Comment:
The `GBINDER_FOURCC` macro has been there since the very beginning, may be considered part of the compile-time API and should remain defined.

Reply:
Ah yes sorry, I thought I was moving that over from the private headers. I think I should undef GBINDER_TRANSACTION here though instead?

Reply:
If `GBINDER_TRANSACTION` and `HIDL_FOURCC` macros are to be defined in the public headers, I suggest to call them e.g. `GBINDER_AIDL_TRANSACTION` and `GBINDER_HIDL_TRANSACTION` to avoid compile-time collisions with other macros with `HIDL_` prefix. And I still think that defining individual transaction codes would be an overkill.

Reply:
Ok. After a bit more digging yesterday I did see that the response to the INTERFACE call just should be interface descriptor string, but that's probably still too high level to handle in gbinder.

Reply:
AIDL-style `INTERFACE` transactions are handled by `gbinder_local_object_interface_transaction()` and HIDL-style `GET_DESCRIPTOR` are handled by `gbinder_local_object_hidl_get_descriptor_transaction()`, are BTW that happens on the looper thread because the object's interface information is immutable and there's nothing to synchronize with the main thread. Are you saying that the response format for those transactions has changed?